File: AtCoder/abc202/D.swift
var arr = readLine()!.split(separator: " ").map({ Int64($0)! })
let N = arr[0] + arr[1]
var ans = ""

var nCr = [[Int64]](repeating: [Int64](repeating: Int64(0), count: 61), count: 61)

nCr[0][0] = Int64(1)

for i in 1 ... 60 {
	nCr[i][0] = Int64(1)
	nCr[i][i] = Int64(1)
	for j in 1 ..< 60 {
		nCr[i][j] = nCr[i - 1][j] + nCr[i - 1][j - 1]
	}
}

for i in Int64(0) ..< N {
	let (i, j) = (Int(arr[0] + arr[1] - 1), Int(arr[1]))
	let cnt = nCr[i][j]
	if cnt >= arr[2] {
		arr[0] -= Int64(1)
		ans += "a"
		
	} else {
		arr[1] -= Int64(1)
		arr[2] -= cnt
		ans += "b"
	}
}

print(ans)

File: AtCoder/abc202/D.kt
private fun readLn() = readLine()!!
private fun readLongs() = readLn().split(" ").map { it.toLong() }

fun main(args: Array<String>) {
    var (a, b, k) = readLongs()
    val dp = Array(61) { Array(61) { 0L } }
    dp[0][0] = 1L
    for (i in 1 until 61) {
        dp[i][0] = 1L
        dp[i][i] = 1L
        for (j in 1 until i) {
            dp[i][j] = dp[i - 1][j] + dp[i - 1][j - 1]
        }
    }
    val n = a + b
    for (i in 0 until n) {
        val cnt = dp[(a + b - 1).toInt()][b.toInt()]
        if (cnt >= k) {
            print("a")
            --a
        } else {
            print("b")
            --b
            k -= cnt
        }
    }
}
